File: Algorithms/Python3.x/714-Best_Time_to_Buy_and_Sell_Stock_with_Transaction_Fee.py
# Runtime: 816 ms, faster than 65.77% of Python3 online submissions for Best Time to Buy and Sell Stock with Transaction Fee.
# Memory Usage: 20.8 MB, less than 12.50% of Python3 online submissions for Best Time to Buy and Sell Stock with Transaction Fee.
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        cash = 0
        hold = -prices[0]
        for i in range(1, len(prices)):
            # sell or hold last day status, hold money
            cash = max(cash, hold + prices[i] - fee)
            # bug or hold last day status, hold stock
            hold = max(hold, cash - prices[i])
        return cash

File: Algorithms/C++/714-Best_Time_to_Buy_and_Sell_Stock_with_Transaction_Fee.cpp
// Runtime: 136 ms, faster than 88.92% of C++ online submissions for Best Time to Buy and Sell Stock with Transaction Fee.
// Memory Usage: 14.9 MB, less than 100.00% of C++ online submissions for Best Time to Buy and Sell Stock with Transaction Fee.
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int cash = 0, hold = -prices[0];
        for (int i = 1; i < prices.size(); ++i) {
            cash = max(cash, hold + prices[i] - fee);
            hold = max(hold, cash - prices[i]);
        }
        return cash;
    }
};
